File: Commons/Reflection/src/main/java/com/greentree/commons/reflection/TypeExtendsFiniteGraph.java
package com.greentree.commons.reflection;

import com.greentree.commons.graph.FiniteGraph;

import java.util.Iterator;
import java.util.stream.StreamSupport;

public record TypeExtendsFiniteGraph(Iterable<Class<?>> classes) implements FiniteGraph<Class<?>> {

    private static final long serialVersionUID = 1L;

    @Override
    public Iterable<? extends Class<?>> getAdjacencyIterable(Object v) {
        var cls = (Class<?>) v;
        var result = StreamSupport.stream(spliterator(), false)
                .filter(x ->
                {
                    if (x.equals(cls.getSuperclass()) || cls.equals(x.getSuperclass()))
                        return true;
                    for (var i : x.getInterfaces())
                        if (i.equals(cls))
                            return true;
                    for (var i : cls.getInterfaces())
                        if (i.equals(x))
                            return true;
                    return false;
                }).toList();
        return result;
    }

    @Override
    public Iterator<Class<?>> iterator() {
        return classes.iterator();
    }

}

File: Commons/Reflection/src/main/java/com/greentree/commons/reflection/TypeExtendsFiniteGraph.kt
package com.greentree.commons.reflection

import com.greentree.commons.graph.FiniteGraph

data class TypeExtendsFiniteGraph(val classes: Iterable<Class<*>>) : FiniteGraph<Class<*>> {

	private fun isAdjacency(cls: Class<*>, parent: Class<*>): Boolean {
		if(cls == parent.superclass || parent == cls.superclass)
			return true
		for(i in cls.interfaces)
			if(i == parent) return true
		for(i in parent.interfaces)
			if(i == cls) return true
		return false
	}

	override fun getAdjacencySequence(cls: Class<*>) = asSequence().filter { isAdjacency(it, cls) }

	override fun iterator(): Iterator<Class<*>> {
		return classes.iterator()
	}
}
